### PR TITLE
Fetch dynamic filters continuously

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

**Comment by sopel39** [comment]
rebased @martint

File: presto-main/src/main/java/io/prestosql/server/remotetask/DynamicFiltersFetcher.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    private final Executor executor;
+    private final HttpClient httpClient;
+    private final RequestErrorTracker errorTracker;
+    private final RemoteTaskStats stats;
+    private final AtomicLong currentRequestStartNanos = new AtomicLong();
+
+    @GuardedBy("this")
+    private long dynamicFiltersVersion = INITIAL_DYNAMIC_FILTERS_VERSION;
+    @GuardedBy("this")
+    private long localDynamicFiltersVersion;
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/server/remotetask/DynamicFiltersFetcher.java:69]: Should this be initialized to `INITIAL_DYNAMIC_FILTERS_VERSION`?

File: presto-main/src/main/java/io/prestosql/server/remotetask/DynamicFiltersFetcher.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+                .setUri(uriBuilderFrom(taskUri).appendPath("dynamicfilters").build())
+                .setHeader(CONTENT_TYPE, JSON_UTF_8.toString())
+                .setHeader(PRESTO_CURRENT_VERSION, Long.toString(localDynamicFiltersVersion))
+                .setHeader(PRESTO_MAX_WAIT, refreshMaxWait.toString())
+                .build();
+
+        errorTracker.startRequest();
+        future = httpClient.executeAsync(request, createFullJsonResponseHandler(dynamicFilterDomainsCodec));
+        currentRequestStartNanos.set(System.nanoTime());
+        Futures.addCallback(future, new SimpleHttpResponseHandler<>(this, request.getUri(), stats), executor);
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/server/remotetask/DynamicFiltersFetcher.java:171]: static import

File: presto-main/src/main/java/io/prestosql/server/remotetask/DynamicFiltersFetcher.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    {
+        if (dynamicFiltersVersion >= newDynamicFiltersVersion) {
+            return;
+        }
+
+        dynamicFiltersVersion = newDynamicFiltersVersion;
+        scheduleNextRequestIfNecessary();
+    }
+
+    private synchronized void scheduleNextRequestIfNecessary()
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/server/remotetask/DynamicFiltersFetcher.java:137]: I would name this "fetchDynamicFilterIfNecessary".. This name is used for the continuous fetchers, so seems a little misleading here.

File: presto-main/src/test/java/io/prestosql/execution/TestDynamicFiltersCollector.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        assertEquals(domains.getDynamicFilterDomains(), ImmutableMap.of(filter, initialDomain));
+
+        // make sure domains are still available when requested again with old version
+        domains = collector.getDynamicFilterDomains(INITIAL_DYNAMIC_FILTERS_VERSION);
+        assertEquals(domains.getVersion(), 1L);
+        assertEquals(domains.getDynamicFilterDomains(), ImmutableMap.of(filter, initialDomain));
+
+        // make sure domains are intersected
+        collector.collectDynamicFilterDomains(ImmutableMap.of(filter, multipleValues(BIGINT, ImmutableList.of(2L))));
+        collector.collectDynamicFilterDomains(ImmutableMap.of(filter, multipleValues(BIGINT, ImmutableList.of(3L))));
```
Comment (dain) [inline presto-main/src/test/java/io/prestosql/execution/TestDynamicFiltersCollector.java:57]: This one of these calls should have a new value in addition to  an existing value.  With this code, the implementation could simply ignore the updates and the test would still work.

**Review by dain — CHANGES_REQUESTED** [review]

**Review by sopel39 — COMMENTED** [review]
ac @dain

File: presto-main/src/main/java/io/prestosql/execution/DynamicFiltersCollector.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    public synchronized long getDynamicFiltersVersion()
+    {
+        return currentVersion;
+    }
+
+    public synchronized VersionedDynamicFilterDomains getDynamicFilterDomains(long callersCurrentVersion)
+    {
+        // Remove dynamic filter domains that are already received by caller.
+        // This assumes there is only one dynamic filters consumer.
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/execution/DynamicFiltersCollector.java:77]: I don't feel strongly, just pointing out the alternative.

**Review by dain — APPROVED** [review]
You might want to squash the three commits together

File: presto-main/src/main/java/io/prestosql/execution/DynamicFiltersCollector.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+        synchronized (this) {
+            long currentVersion = ++this.currentVersion;
+            for (Map.Entry<DynamicFilterId, Domain> entry : newDynamicFilterDomains.entrySet()) {
+                dynamicFilterDomains.merge(
+                        entry.getKey(),
+                        new VersionedDomain(currentVersion, entry.getValue().simplify()),
+                        (oldDomain, newDomain) -> new VersionedDomain(
+                                max(oldDomain.getVersion(), newDomain.getVersion()),
+                                oldDomain.getDomain().intersect(newDomain.getDomain())));
```
Comment (rohangarg) [inline presto-main/src/main/java/io/prestosql/execution/DynamicFiltersCollector.java:62]: should there be `simplify` after intersect too?

Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/execution/DynamicFiltersCollector.java:62]: > should there be simplify after intersect too?

Currently there is only one (final) DF version.

**Pull request merged**